incipyt/_internal/utils.py

```diff
@@ -0,0 +1,143 @@
+import collections.abc
+from string import Formatter
+
+import click
+
+
+class Requires:
+    def __init__(self, template, confirmed=False, sanitizer=None, **kwargs):
+        self._confirmed = confirmed
+        self._sanitizer = sanitizer
+        self._template = template
+        self._kwargs = kwargs
+
+    def __str__(self):
+        return f"f'{self._template}'"
+
+    def __call__(self, environment):
+        for _, key, _, _ in Formatter().parse(self._template):
+            if key not in self._kwargs:
+                continue
+
+            environment.push(key, self._kwargs[key], confirmed=self._confirmed)
+
+        args = {
+            key: (
+                self._sanitizer(key, environment.pull(key))
+                if self._sanitizer
+                else environment.pull(key)
+            )
+            for _, key, _, _ in Formatter().parse(self._template)
+            if key is not None
+        }
+        if any(not v for v in args.values()):
+            return None
+
+        return self._template.format(**args)
+
+
+class MultipleValues:
+    def __init__(self, head, tail):
+        self._values = (
+            [head] + tail._values if isinstance(tail, MultipleValues) else [head, tail]
+        )
+
+    def __call__(self, environment):
+        return click.prompt(
+            "Conflicting configuration, choose between",
+            type=click.Choice(
+                [
+                    value(environment) if callable(value) else value
+                    for value in self._values
+                ]
+            ),
+        )
+
+    def __repr__(self):
+        return f"{type(self).__name__}({self._values})"
+
+    def __eq__(self, other):
+        try:
+            return self._values == other._values
+        except AttributeError:
+            return False
+
+
+def append_unique(config, value):
+    assert isinstance(config, collections.abc.MutableSequence)
+
+    if value not in config:
+        config.append(value)
+
+
+def is_not_string_sequence(obj):
+    return (
+        isinstance(obj, collections.abc.Sequence)
+        and not isinstance(obj, collections.abc.ByteString)
+        and not isinstance(obj, str)
+    )
+
+
+class TemplateDict(collections.UserDict):
+    def __init__(self, mapping=None):
+        # If an existing mapping is provided, this class will act like a proxy
+        # around it
+        self.data = {} if mapping is None else mapping
```

Comment:
Why are you dealing with the "non proxy" case ?

I don't think we are going to use it in such a case.

Reply:
Mainly for tesing, instancing with `TemplateDict()` seems more natural than `TemplateDict({})`